### PR TITLE
NOISSUE - Refactor mProxy

### PR DESCRIPTION
**Pull request by dusanb94**
Refactor mProxy. Close both connections on streaming errors.

### CODE REVIEW

File: examples/simple/simple.go [hunk 1]
```diff
@@ -8,62 +8,62 @@ import (
...
+func New(logger logger.Logger) *Handler {
+	return &Handler{
 		logger: logger,
 	}
 }
 
 // AuthConnect is called on device connection,
 // prior forwarding to the MQTT broker
-func (e *Event) AuthConnect(c *session.Client) error {
+func (e *Handler) AuthConnect(c *session.Client) error {
```
Comment (manuio) [inline examples/simple/simple.go:27]: `h *Handler`

File: pkg/session/session.go [hunk 2]
```diff
@@ -32,7 +32,7 @@ func New(inbound, outbound net.Conn, event EventHandler, logger logger.Logger) *
 		logger:   logger,
 		inbound:  inbound,
 		outbound: outbound,
-		event:    event,
+		handler:  event,
```
Comment (manuio) [inline pkg/session/session.go:35]: maybe `handler:  handler,`

File: pkg/session/events.go [hunk 3]
```diff
@@ -1,7 +1,7 @@
 package session
 
-// Event is an interface for mProxy hooks
-type Event interface {
+// EventHandler is an interface for mProxy hooks
+type EventHandler interface {
```
Comment (manuio) [inline pkg/session/events.go:4]: I would use `Handler` and also rename the file

**Comment by manuio** [comment]
@dusanb94 I don't see any explicit connection closing, it's normal?

File: examples/simple/simple.go [hunk 4]
```diff
@@ -8,62 +8,62 @@ import (
...
 		logger: logger,
 	}
 }
 
 // AuthConnect is called on device connection,
 // prior forwarding to the MQTT broker
-func (e *Event) AuthConnect(c *session.Client) error {
-	e.logger.Info(fmt.Sprintf("AuthRegister() - clientID: %s, username: %s, password: %s", c.ID, c.Username, string(c.Password)))
+func (h *Handler) AuthConnect(c *session.Client) error {
+	h.logger.Info(fmt.Sprintf("AuthRegister() - clientID: %s, username: %s, password: %s", c.ID, c.Username, string(c.Password)))
```
Comment (manuio) [inline examples/simple/simple.go:28]: Do we need format `AuthRegister() -` in logs?

File: examples/simple/simple.go [hunk 5]
```diff
@@ -8,62 +8,62 @@ import (
...
-	e.logger.Info(fmt.Sprintf("Unsubscribe() - username: %s, clientID: %s, topics: %s", c.Username, c.ID, strings.Join(*topics, ",")))
+func (h *Handler) Unsubscribe(c *session.Client, topics *[]string) {
+	h.logger.Info(fmt.Sprintf("Unsubscribe() - username: %s, clientID: %s, topics: %s", c.Username, c.ID, strings.Join(*topics, ",")))
 }
 
 // Disconnect on conection lost
-func (e *Event) Disconnect(c *session.Client) {
-	e.logger.Info(fmt.Sprintf("Disconnect() - client with username: %s and ID: %s disconenectd", c.Username, c.ID))
+func (h *Handler) Disconnect(c *session.Client) {
+	h.logger.Info(fmt.Sprintf("Disconnect() - client with username: %s and ID: %s disconenectd", c.Username, c.ID))
```
Comment (manuio) [inline examples/simple/simple.go:68]: Fix `disconenectd`

File: pkg/session/session.go [hunk 6]
```diff
@@ -19,32 +22,50 @@ type Session struct {
...
 
 	go s.stream(up, s.inbound, s.outbound, errs)
 	go s.stream(down, s.outbound, s.inbound, errs)
 
-	err := <-errs
-	s.event.Disconnect(&s.Client)
-	return err
+	err1 := <-errs
+
+	// Close both connections when error occurs.
```
Comment (drasko) [inline pkg/session/session.go:50]: I actually preferred defer implementation. Connections are opened on upper level - they should be closed there.

File: pkg/session/handler.go [hunk 7]
```diff
@@ -1,7 +1,7 @@
 package session
 
-// Event is an interface for mProxy hooks
-type Event interface {
+// Handler is an interface for mProxy hooks
+type Handler interface {
```
Comment (drasko) [inline pkg/session/handler.go:4]: Would Emmiter be better name maybe?

Comment (dusanb94) [reply pkg/session/handler.go:4]: Let's stick with `Handler` for now, it's generic enough.

File: examples/simple/simple.go (hunk 4 above)
Comment (dusanb94) [reply examples/simple/simple.go:28]: Not really, but this is only an example.

Comment (nmarcetic) [reply examples/simple/simple.go:28]: @manuio @dusanb94 Leave this logging pls. It's priceless for debugging and you can control it with a log level in different env. Plus its example yes :)

**Pull request merged**